crates/sui-move-build/src/linters/share_owned.rs

```diff
@@ -0,0 +1,267 @@
+        let ExecutionContext { diags } = context;
+        diags
+    }
+
+    fn exp_custom(
+        &self,
+        context: &mut ExecutionContext,
+        state: &mut State,
+        e: &Exp,
+    ) -> Option<Vec<Value>> {
+        use move_compiler::hlir::ast::UnannotatedExp_ as E;
+
+        if let E::Pack(_, _, fields) = &e.exp.value {
+            for (_, _, inner) in fields.iter() {
+                self.exp(context, state, inner);
+            }
+            return Some(vec![Value::FreshObj]);
+        };
+
+        None
+    }
+
+    fn call_custom(
+        &self,
+        context: &mut ExecutionContext,
+        _state: &mut State,
+        loc: &Loc,
+        return_ty: &Type,
+        f: &ModuleCall,
+        args: Vec<Value>,
+    ) -> Option<Vec<Value>> {
+        if SHARE_FUNCTIONS
+            .iter()
+            .any(|(addr, module, fun)| f.is(addr, module, fun))
+            && args[0] != Value::FreshObj
+        {
+            let msg = "You may be trying to share an owned object.";
+            let uid_msg = "This argument should in most cases come from an object created within the same function.";
+            let mut d = diag!(
+                SHARE_OWNED_DIAG,
+                (*loc, msg),
+                (f.arguments.exp.loc, uid_msg)
+            );
+            if let Value::NotFreshObj(l) = args[0] {
+                d.add_secondary_label((l, "A potentially shared object coming from here"))
+            }
+            context.add_diag(d)
+        }
+        Some(match &return_ty.value {
+            Type_::Unit => vec![],
+            Type_::Single(t) => {
+                let v = if is_obj_type(t) {
+                    Value::NotFreshObj(t.loc)
+                } else {
+                    Value::Other
+                };
+                vec![v]
+            }
+            Type_::Multiple(types) => types
+                .iter()
+                .map(|t| {
+                    if is_obj_type(t) {
+                        Value::NotFreshObj(t.loc)
+                    } else {
+                        Value::Other
+                    }
+                })
+                .collect(),
+        })
+    }
+
+    fn command_custom(&self, _: &mut ExecutionContext, _: &mut State, _: &Command) -> bool {
+        false
+    }
+
+    fn lvalue_custom(
+        &self,
+        context: &mut ExecutionContext,
+        state: &mut State,
+        l: &LValue,
+        _value: &Value,
+    ) -> bool {
+        use LValue_ as L;
+
+        let sp!(_, l_) = l;
+        if let L::Unpack(_, _, fields) = l_ {
+            for (f, l) in fields {
+                let v = if is_obj(l) {
+                    <Self::State as SimpleDomain>::Value::NotFreshObj(f.loc())
+                } else {
+                    <Self::State as SimpleDomain>::Value::default()
+                };
+                self.lvalue(context, state, l, v)
+            }
+            return true;
```

Comment:
thoughts on the bool thing? (where true is skip and false is continue)

I came up with it and I find it a bit confusing reading it lol

Reply:
I didn't give it that much thought but I found it pretty sensible to work with.